### PR TITLE
Fix baseUrl origin and support serverUrl and apiUrl are different

### PR DESCRIPTION
**Pull request by hoangdat**
Cc @tnstung @quantranhong1999

### CODE REVIEW

File: core/test/presentation/extensions/uri_extension_test.dart [hunk 1]
```diff
@@ -55,5 +55,45 @@ void main() {
 
       expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
     });
+
+    test('convertToQualifiedUrl() should return qualified url when baseUrl is `https://domain.com:2000/jmap` and sourceUrl is `https://domain.com:2001/jmap`', () async {
+      final baseUrl =  Uri.parse('https://domain.com:2000/jmap');
+      final sourceUrl =  Uri.parse('https://domain.com:2001/jmap');
```
Comment (dab246) [inline core/test/presentation/extensions/uri_extension_test.dart:61]:
```suggestion
      final baseUrl = Uri.parse('https://domain.com:2000/jmap');
      final sourceUrl = Uri.parse('https://domain.com:2001/jmap');
```

File: core/test/presentation/extensions/uri_extension_test.dart [hunk 2]
```diff
@@ -55,5 +55,45 @@ void main() {
...
+
+      final qualifiedUrlExpected = Uri.parse('https://domain.com:2001/jmap');
+      final qualifiedUrlResult = sourceUrl.toQualifiedUrl(baseUrl: baseUrl);
+
+      expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
+    });
+
+    test('convertToQualifiedUrl() should return qualified url when baseUrl is `https://domain.com:2000/jmap` and sourceUrl is `https://domain.com:2001/jmap/`', () async {
+      final baseUrl =  Uri.parse('https://domain.com:2000/jmap');
+      final sourceUrl =  Uri.parse('https://domain.com:2001/jmap/');
```
Comment (dab246) [inline core/test/presentation/extensions/uri_extension_test.dart:71]: idem

File: core/test/presentation/extensions/uri_extension_test.dart [hunk 3]
```diff
@@ -55,5 +55,45 @@ void main() {
...
+
+      final qualifiedUrlExpected = Uri.parse('https://domain.com:2001/jmap');
+      final qualifiedUrlResult = sourceUrl.toQualifiedUrl(baseUrl: baseUrl);
+
+      expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
+    });
+
+    test('convertToQualifiedUrl() should return qualified url when baseUrl is `https://domain.com:20001/jmap` and sourceUrl is `https://domain.com:2001/jmap`', () async {
+      final baseUrl =  Uri.parse('https://domain.com:2001/jmap');
+      final sourceUrl =  Uri.parse('https://domain.com:2001/jmap');
```
Comment (dab246) [inline core/test/presentation/extensions/uri_extension_test.dart:81]: idem

File: core/test/presentation/extensions/uri_extension_test.dart [hunk 4]
```diff
@@ -55,5 +55,45 @@ void main() {
...
+
+      final qualifiedUrlExpected = Uri.parse('https://domain.com:2001/jmap');
+      final qualifiedUrlResult = sourceUrl.toQualifiedUrl(baseUrl: baseUrl);
+
+      expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
+    });
+
+    test('convertToQualifiedUrl() should return qualified url when baseUrl is localhost and sourceUrl is `https://domain.com:2001/basiauth/jmap`', () async {
+      final baseUrl =  Uri.parse('https://localhost:9080/basicauth');
+      final sourceUrl =  Uri.parse('https://domain.com:2001/basicauth/jmap');
```
Comment (dab246) [inline core/test/presentation/extensions/uri_extension_test.dart:91]: idem

File: test/features/login/data/repository/credential_repository_imp_test.dart [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+@GenerateMocks([AuthenticationInfoCacheManager])
+void main() {
+  late SharedPreferences sharedPreferences;
+  late AuthenticationInfoCacheManager authenticationInfoCacheManager;
+  late CredentialRepositoryImpl credentialRepositoryImpl;
+
+  setUp(() async {
+    SharedPreferences.setMockInitialValues({});
+    sharedPreferences = await SharedPreferences.getInstance();
```
Comment (dab246) [inline test/features/login/data/repository/credential_repository_imp_test.dart:17]: IMO, It has to be initialized before `setMockInitialValues ` right?

File: test/features/login/data/repository/credential_repository_imp_test.dart [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    credentialRepositoryImpl = CredentialRepositoryImpl(sharedPreferences, authenticationInfoCacheManager);
+  });
+
+  group('test saveBaseUrl', () {
+    test('baseUrl should be save exactly when url has no sub path', () async {
+      final url = Uri.parse('https://example.com');
+      credentialRepositoryImpl.saveBaseUrl(url);
+
+      final storedUrl = await credentialRepositoryImpl.getBaseUrl();
+      expect(storedUrl, url);
```
Comment (dab246) [inline test/features/login/data/repository/credential_repository_imp_test.dart:28]:
```suggestion
      expect(storedUrl, equal(url));
```

File: test/features/login/data/repository/credential_repository_imp_test.dart [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    });
+
+    test('baseUrl should be save exactly when url has sub path', () async {
+      final url = Uri.parse('https://example.com/basicauth');
+      credentialRepositoryImpl.saveBaseUrl(url);
+
+      final storedUrl = await credentialRepositoryImpl.getBaseUrl();
+      expect(storedUrl, url);
```
Comment (dab246) [inline test/features/login/data/repository/credential_repository_imp_test.dart:36]: idem

File: test/features/login/data/repository/credential_repository_imp_test.dart [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    });
+
+    test('baseUrl should be save exactly when url has sub port', () async {
+      final url = Uri.parse('https://example.com:8080/');
+      credentialRepositoryImpl.saveBaseUrl(url);
+
+      final storedUrl = await credentialRepositoryImpl.getBaseUrl();
+      expect(storedUrl, url);
```
Comment (dab246) [inline test/features/login/data/repository/credential_repository_imp_test.dart:44]: idem

File: test/features/login/data/repository/credential_repository_imp_test.dart [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    });
+
+    test('baseUrl should be save exactly when url has port and sub path', () async {
+      final url = Uri.parse('https://example.com:8080/basicauth');
+      credentialRepositoryImpl.saveBaseUrl(url);
+
+      final storedUrl = await credentialRepositoryImpl.getBaseUrl();
+      expect(storedUrl, url);
```
Comment (dab246) [inline test/features/login/data/repository/credential_repository_imp_test.dart:52]: idem

File: test/features/login/data/repository/credential_repository_imp_test.dart [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    });
+
+    test('baseUrl should be save exactly when url has multiple sub paths', () async {
+      final url = Uri.parse('https://example.com:8080/basicauth/jmap');
+      credentialRepositoryImpl.saveBaseUrl(url);
+
+      final storedUrl = await credentialRepositoryImpl.getBaseUrl();
+      expect(storedUrl, url);
```
Comment (dab246) [inline test/features/login/data/repository/credential_repository_imp_test.dart:60]: idem

File: test/features/login/data/repository/credential_repository_imp_test.dart (hunk 5 above)
Comment (hoangdat) [reply test/features/login/data/repository/credential_repository_imp_test.dart:17]: If we call it after getInstance. Error:
image

`SharedPreferences` is platform based class, so we need to set `initMock` before create this instance

File: core/lib/presentation/extensions/uri_extension.dart [hunk 11]
```diff
@@ -10,12 +10,27 @@ extension URIExtension on Uri {
...
-      final qualifiedUrl = baseUrlValid + sourceUrlValid;
-      log('SessionUtils::toQualifiedUrl():qualifiedUrl: $qualifiedUrl');
-      return Uri.parse(qualifiedUrl);
+      if (!hasOrigin) {
+        final baseUrlValid = baseUrl.toString().removeLastSlashOfUrl();
+        final sourceUrlValid = toString().addFirstSlashOfUrl().removeLastSlashOfUrl();
+        log('SessionUtils::toQualifiedUrl():baseUrlValid: $baseUrlValid | sourceUrlValid: $sourceUrlValid');
+        final qualifiedUrl = baseUrlValid + sourceUrlValid;
+        log('SessionUtils::toQualifiedUrl():qualifiedUrl: $qualifiedUrl');
+        return Uri.parse(qualifiedUrl);
```
Comment (tddang-linagora) [inline core/lib/presentation/extensions/uri_extension.dart:19]: Please unit test this case

Comment (hoangdat) [reply core/lib/presentation/extensions/uri_extension.dart:19]: we had these:
```dart
    test('convertToQualifiedUrl() should return qualified url when baseUrl is `https://domain.com` and sourceUrl is `/jmap/`', () async {
      final baseUrl = Uri.parse('https://domain.com');
      final sourceUrl = Uri.parse('/jmap/');

      final qualifiedUrlExpected = Uri.parse('https://domain.com/jmap');
      final qualifiedUrlResult = sourceUrl.toQualifiedUrl(baseUrl: baseUrl);

      expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
    });

    test('convertToQualifiedUrl() should return qualified url when baseUrl is `https://domain.com` and sourceUrl is `/jmap`', () async {
      final baseUrl = Uri.parse('https://domain.com');
      final sourceUrl = Uri.parse('/jmap');

      final qualifiedUrlExpected = Uri.parse('https://domain.com/jmap');
      final qualifiedUrlResult = sourceUrl.toQualifiedUrl(baseUrl: baseUrl);

      expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
    });

    test('convertToQualifiedUrl() should return qualified url when baseUrl is `https://domain.com/jmap` and sourceUrl is empty', () async {
      final baseUrl = Uri.parse('https://domain.com/jmap');
      final sourceUrl = Uri.parse('');

      final qualifiedUrlExpected = Uri.parse('https://domain.com/jmap');
      final qualifiedUrlResult = sourceUrl.toQualifiedUrl(baseUrl: baseUrl);

      expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
    });

    test('convertToQualifiedUrl() should return qualified url when baseUrl is `https://domain.com/jmap` and sourceUrl is `/`', () async {
      final baseUrl = Uri.parse('https://domain.com/jmap');
      final sourceUrl = Uri.parse('/');

      final qualifiedUrlExpected = Uri.parse('https://domain.com/jmap');
      final qualifiedUrlResult = sourceUrl.toQualifiedUrl(baseUrl: baseUrl);

      expect(qualifiedUrlResult, equals(qualifiedUrlExpected));
    });
```

**Review by tddang-linagora — APPROVED** [review]
lgtm

**Review by dab246 — APPROVED** [review]
- [x] Please rebase it

**Pull request merged**